src/Chapter2.hs

```diff
@@ -605,7 +637,12 @@ Implement a function that duplicates each element of the list
 
 -}
 duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
+duplicate l = go True l []
+  where
+    go :: Bool -> [a] -> [a] -> [a]
+    go _ [] acc = acc
+    go False (x:xs) acc = go True xs (acc ++ [x])
+    go True (x:xs) acc = go False (x:xs) (acc ++ [x])
```

Comment:
Usually, adding elements to the end of the list is costly. Your implementation is correct :+1: But you can try to implement this function more efficiently by using `:` instead `++`. If you can think, how to avoid adding elements to the end of the list, and instead add to the beginning, but get the same result, that would be amazing!

In general, it's always a good idea to try to replace `l ++ [x]` with `x : l` whenever possible, though, in some cases it might not be that trivial to achieve the same result. Remember the analogy with trains :bullettrain_side: :train: :train: :train:

Also, I see that you have a `Bool` variable `acc`, but it is not actually used in this implementation. I guess, it was copy-pasted from `takeEven`, but here you can remove it :slightly_smiling_face:

Reply:
Yeah I had initially naively used `x:l` and then was surprised to discover that the list was the reverse of what I wanted! 

I would assume that reversing the list after is still a pretty costly operation. Can you tell me a bit more about the time complexity in that case? Is (:) + reverse better than (++)? Intuitively, I would assume that one is O(n) and the other is O(n^2).

As to the Bool variable, I think it does serve the purpose of not infinitely repeating any one element. It is kind of a `shouldRepeat` variable, and then on the next step `shouldRepeat` becomes `NOT shouldRepeat` and the current element is dropped.

Reply:
`reverse` has linear complexity, but that's not the most problematic thing with `reverse`. `(:)` + `reverse` is better than `++` of a single element to the end of the list, and your complexity analysis is correct. But `(:)` without `reverse` is not even better. `reverse` needs to walk list to the end to return the result. But with lazy evaluation in Haskell, it's always desirable to output elements immediately, in a streaming fashion. Because of lazy evaluation, if functions are implemented memory, you get streaming for free, constant memory usage, automatic work with infinite lists, and everything composes at the same time.